src/main/java/io/oneko/project/ProjectVersion.java

```diff
@@ -83,20 +85,40 @@ default String safeVersionName() {
 		return safeName.substring(0, Math.min(safeName.length(), 63));
 	}
 
+	default String[] getCalculatedUrls() {
+		final Map<String, String> implicitTemplateVariables = getImplicitTemplateVariables();
+		TemplateContext context = new TemplateContext();
+		implicitTemplateVariables.forEach(context::set);
+		context.set("fn", TemplateFunctions.class);
+		List<String> urlTemplates = new ArrayList<>();
+		if (getProject() != null) {
+			urlTemplates.addAll(getProject().getUrlTemplates());
+		}
+		urlTemplates.addAll(getUrlTemplates());
+		return urlTemplates.stream()
+				.map(url -> {
+					TemplateLoader.MapTemplateLoader loader = new TemplateLoader.MapTemplateLoader();
+					loader.set("tpl", url);
+					final Template tpl = loader.load("tpl");
+					return tpl.render(context);
+				}).toArray(String[]::new);
+	}
+
 	/**
 	 * Provides a mutable copy of all template variables retrieved by merging the ones from this version's
 	 * {@link #getProject()} and this version's own {@link #getTemplateVariables()}.
 	 *
 	 * @return Never <code>null</code>
 	 */
-	default Map<String, String> calculateEffectiveTemplateVariables() {
-		Map<String, String> mergedTemplateVariables = new HashMap<>();
+	default Map<String, Object> calculateEffectiveTemplateVariables() {
```

Comment:
> We should introduce a new interface instead of relying on `Object`. Maybe `TemplateVariableValue`?

Ideally I'd make it `Map<String, String|String[]>` but that's too TypeScript-y for Java. 🙃

> Even better: we just add the urls one by one with `key + "[" + i + "]"`.

This does not seem like a good idea, to be honest. We're getting this for free already just by using the array.

Reply:
We agreed on leaving this as it is.